Mathlib/CategoryTheory/Bicategory/Extension.lean

```diff
@@ -0,0 +1,147 @@
+/-
+Copyright (c) 2023 Yuma Mizuno. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yuma Mizuno
+-/
+import Mathlib.CategoryTheory.Bicategory.Basic
+
+/-!
+# Extensions and lifts in bicategories
+
+We introduce the concept of extensions and lifts within the bicategorical framework. These concepts
+are defined by commutative diagrams in the (1-)categorical context. Within the bicategorical
+framework, commutative diagrams are replaced by 2-morphisms. Depending on the orientation of the
+2-morphisms, we define both left and right extensions (likewise for lifts).
+
+The use of left and right here is a common one in the theory of Kan extensions.
+
+## References
+* https://ncatlab.org/nlab/show/lifts+and+extensions
+* https://ncatlab.org/nlab/show/Kan+extension
+
+## Todo
+API for left lifts, right extensions, and right lifts
+-/
+
+namespace CategoryTheory
+
+namespace Bicategory
+
+universe w v u
+
+variable {B : Type u} [Bicategory.{w, v} B] {a b c : B}
+
+/-- Triangle diagrams for (left) extensions.
+```
+  b
+  △ \
+  |   \ extension  △
+f |     \          | unit
+  |       ◿
+  a - - - ▷ c
+      g
+```
+-/
+structure LeftExtension (f : a ⟶ b) (g : a ⟶ c) where
+  /-- The extension of `g` along `f`. -/
+  extension : b ⟶ c
+  /-- The 2-morphism filling the triangle diagram. -/
+  unit : g ⟶ f ≫ extension
```

Comment:
If we introduce the whiskering functors, the category `LeftExtension` identifies to a suitable category `StructuredArrow`. Recent PRs by @TwoFX like https://github.com/leanprover-community/mathlib4/pull/6248 contain useful results about these categories. I know that this type of results about `(Co)structuredArrow` categories will be very useful when proving things about Kan extensions (e.g. total derived functors for which I have done some experiments). Then, again, in order to avoid duplication of code, it would be great to base the definition of `Left/RightExtension` using this so that we can later use the existing and future developments of the `StructuredArrow` API. Still, you may introduce more convenient specialized abbreviations like "LeftExtension.extension" or "LeftExtension.unit" instead of `Comma.right` and `Comma.hom`.
```lean
@[simps]
def whiskeringLeft (a b c : B) : (a ⟶ b) ⥤ (b ⟶ c) ⥤ (a ⟶ c) where
  obj f :=
    { obj := fun e => f ≫ e
      map := fun φ => f ◁ φ }
  map ψ :=
    { app := fun e => ψ ▷ e
      naturality := fun {e₁ e₂} φ => whisker_exchange _ _ }

def LeftExtension (f : a ⟶ b) (g : a ⟶ c) :=
  StructuredArrow g ((whiskeringLeft a b c).obj f)
```